pkg/apis/pipeline/v1beta1/when_types.go

```diff
@@ -0,0 +1,139 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"k8s.io/apimachinery/pkg/selection"
+)
+
+// WhenExpression allows a PipelineTask to declare expressions to be evaluated before the Task is run
+// to determine whether the Task should be executed or skipped
+type WhenExpression struct {
+	// Input is the string for guard checking which can be a static input or an output from a parent Task
+	Input string `json:"input,omitempty"`
+	// Operator that represents an Input's relationship to the values
+	Operator selection.Operator `json:"operator,omitempty"`
+	// Values is an array of strings, which is compared against the input, for guard checking
+	// It must be non-empty
+	Values []string `json:"values,omitempty"`
+}
+
+// GetVarSubstitutionExpressionsForWhenExpression extracts all the values between "$(" and ")"" for a when expression
+func (we *WhenExpression) GetVarSubstitutionExpressionsForWhenExpression() ([]string, bool) {
+	var allExpressions []string
+	allExpressions = append(allExpressions, validateString(we.Input)...)
+	for _, value := range we.Values {
+		allExpressions = append(allExpressions, validateString(value)...)
+	}
+	return allExpressions, len(allExpressions) != 0
+}
+
+func (we *WhenExpression) valuesContainsInput() bool {
+	for i := range we.Values {
+		if we.Values[i] == we.Input {
+			return true
+		}
+	}
+	return false
+}
+
+func (we *WhenExpression) isTrue() bool {
+	if we.Operator == selection.In {
+		return we.valuesContainsInput()
+	}
+	return !we.valuesContainsInput()
+}
+
+func (we *WhenExpression) hasVariable() bool {
+	if _, hasVariable := we.GetVarSubstitutionExpressionsForWhenExpression(); hasVariable {
+		return true
+	}
+	return false
+}
+
+func (we *WhenExpression) applyReplacements(replacements map[string]string) WhenExpression {
+	replacedInput := ApplyReplacements(we.Input, replacements)
+
+	var replacedValues []string
+	for _, val := range we.Values {
+		replacedValues = append(replacedValues, ApplyReplacements(val, replacements))
+	}
+
+	return WhenExpression{Input: replacedInput, Operator: we.Operator, Values: replacedValues}
+}
+
+// WhenExpressions is an array of When Expressions which are used to specify whether a Task should be executed or skipped.
+type WhenExpressions []WhenExpression
+
+// AreTrue indicates whether the When Expression are true.
+func (wes WhenExpressions) AreTrue() bool {
+	for _, we := range wes {
+		if isTrue := we.isTrue(); !isTrue {
```

Comment:
NIT: I don't think `isTrue` is required here, this `if` condition can be simplified to:

```
if !we.isTrue() {
  return false
}
```

And similar `if` condition on line 95.